airflow/operators/subdag_operator.py

```diff
@@ -89,14 +88,71 @@ def __init__(
                         )
                     )
 
-        self.subdag = subdag
-        # Airflow pool is not honored by SubDagOperator.
-        # Hence resources could be consumed by SubdagOperators
-        # Use other executor with your own risk.
-        self.executor = executor
-
-    def execute(self, context):
-        ed = context['execution_date']
-        self.subdag.run(
-            start_date=ed, end_date=ed, donot_pickle=True,
-            executor=self.executor)
+    def _get_dagrun(self, execution_date):
+        dag_runs = DagRun.find(
+            dag_id=self.subdag.dag_id,
+            execution_date=execution_date,
+        )
+        return dag_runs[0] if dag_runs else None
+
+    @provide_session
+    def _reset_dag_run_and_task_instances(self, dag_run, execution_date, session=None):
```

Comment:
Instead of using the `provide_session` here, we could also use the `create_session`, the rollback logic is already there: https://github.com/apache/airflow/blob/master/airflow/utils/db.py#L32-L45

This would translate to:
```
        with create_session() as session:
            dag_run.state = State.RUNNING
            session.merge(dag_run)
            failed_task_instances = (
                session
                .query(TaskInstance)
                .filter(TaskInstance.dag_id == self.subdag.dag_id)
                .filter(TaskInstance.execution_date == execution_date)
                .filter(TaskInstance.state.in_([State.FAILED, State.UPSTREAM_FAILED]))
            )

            for task_instance in failed_task_instances:
                task_instance.state = State.NONE
                session.merge(task_instance)
```
WDYT?

Reply:
Yeah this is better as it wraps the `try ... finally rollback` logic in the context manager.